python/tests/test_user_session.py

```diff
@@ -0,0 +1,79 @@
+import base64
+import json
+from unittest import mock
+
+import pytest
+
+from numerous._client._graphql_client import GraphQLClient
+from numerous.generated.graphql.client import Client as GQLClient
+from numerous.user import User
+from numerous.user_session import Session
+
+
+class CookieGetterStub:
+    def __init__(self, cookies: dict[str, str]) -> None:
+        self._cookies = cookies
+
+    def cookies(self) -> dict[str, str]:
+        """Get the cookies associated with the current request."""
+        return self._cookies
+
+
+ORGANIZATION_ID = "test_org"
+
+
+@pytest.fixture(autouse=True)
+def _set_env_vars(monkeypatch: pytest.MonkeyPatch) -> None:
+    monkeypatch.setenv("NUMEROUS_API_URL", "url_value")
+    monkeypatch.setenv("NUMEROUS_ORGANIZATION_ID", ORGANIZATION_ID)
+    monkeypatch.setenv("NUMEROUS_API_ACCESS_TOKEN", "token")
+
+
+@pytest.fixture
+def mock_gql_client() -> GQLClient:
+    return mock.Mock(GQLClient)
+
+
+@pytest.fixture
+def mock_graphql_client(mock_gql_client: GQLClient) -> GraphQLClient:
+    return GraphQLClient(mock_gql_client)
+
+
+def test_user_property_raises_value_error_when_no_cookie(
+    mock_graphql_client: GraphQLClient,
+) -> None:
+    cg = CookieGetterStub({})
+    session = Session(cg, _client=mock_graphql_client)
+    with pytest.raises(
+        ValueError, match="Invalid user info in cookie or cookie is missing"
+    ):
+        # ruff: noqa: B018
+        session.user
+
+
+def test_user_property_returns_user_when_valid_cookie(
+    mock_graphql_client: GraphQLClient,
+) -> None:
+    user_info = {"user_id": "1", "name": "Test User"}
+    encoded_info = base64.b64encode(json.dumps(user_info).encode()).decode()
+    cg = CookieGetterStub({"numerous_user_info": encoded_info})
+
+    session = Session(cg, _client=mock_graphql_client)
+    assert isinstance(session.user, User)
+    assert session.user is not None
+    assert session.user.id == "1"
+    assert session.user.name == "Test User"
+
+
+def test_user_info_returns_decoded_info_for_valid_cookie(
+    mock_graphql_client: GraphQLClient,
+) -> None:
+    user_info = {"user_id": "1", "name": "Test User"}
+    encoded_info = base64.b64encode(json.dumps(user_info).encode("utf-8")).decode(
+        "utf-8"
+    )
+    cg = CookieGetterStub({"numerous_user_info": encoded_info})
+    session = Session(cg, _client=mock_graphql_client)
+    assert session.user is not None
+    assert session.user.id == "1"
+    assert session.user.name == "Test User"
```

Comment:
These two test cases seem to do be duplicates? Also you still need to separate "arrange/act/assert" stages of the tests, e.g. put an empty line before all asserts, and after all the setup :smile:

Reply:
The separation is a general thing, not just for this selected range :smile:

Reply:
@Lasse-numerous  

1. You removed the test which had fixed the issue with the type which is not needed since we do static type checking. So remove the `assert isintance(...)`

2. You misunderstood me. The point is to separate the three stages of a test: (1) arrange (or setup), (2) act (running actual code under test), and (3) assert (checking stuff).
Instead you added separation between each assertion, which we do not really want.

For example:

```python
def test_user_property_returns_user_when_valid_cookie(
    mock_graphql_client: GraphQLClient,
) -> None:
    user_info = {"user_id": "1", "name": "Test User"}
    encoded_info = base64.b64encode(json.dumps(user_info).encode()).decode()
    cg = CookieGetterStub({"numerous_user_info": encoded_info})
    # separation after setup

    session = Session(cg, _client=mock_graphql_client)
    # separation before assertions

    assert session.user is not None
    assert session.user.id == "1"
    assert session.user.name == "Test User"
```

Reply:
Hmm, ok. I have removed the isinstance check and arranged the tests for arrange, act, and assert as I understood it.